.github/workflows/push.yml

```diff
@@ -191,6 +191,78 @@ jobs:
           runTestsParameters: >-
             -d zend_extension=opcache.so
             -d opcache.enable_cli=1
+  ALPINE:
+    if: github.repository == 'php/php-src' || github.event_name == 'pull_request'
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - debug: false
+            zts: false
+            asan: false
+          - debug: true
+            zts: true
+            asan: true
```

Comment:
I would suggest just the latter, two builds seem like overkill to me. Instead, you could add both to nightly.

Reply:
> Honestly, I wonder if it wouldn't be more useful to just move this to nightly but with all extensions enabled. WDYT? Or do you think this tests something unique (apart from alpine/musl)?

I planned to also add a nightly job after this, but I don't mind moving this to nightly initially.

Reply:
I won't object to either, but IMO the push itself is not a big value-add if it's unlikely to break, without the other builds breaking.

Reply:
I would prefer to have at least one Alpine test on push as I almost never check nightly and it's also harder to test the possible fixes. Ideally one ZTS non ASAN for as many extensions as possible would be great.

Reply:
Somebody must be looking at nightly daily anyway, which I am. My main thinking is: _If_ musl works with few changes before any CI, it doesn't seem highly likely to find new error on push. If it does in nightly, is that not enough? Anyway,

Reply:
I've re-added the push job in the meantime (zts non-asan). What would be the drawbacks of having a push job?

Reply:
What about removing "LINUX_X32_DEBUG_ZTS" from push and replacing it with Alpine?

Reply:
> What would be the drawbacks of having a push job?

Possibly increasing CI feedback delay (depending on how slow the job is) and burning the planet using the additional energy consumed.

Reply:
32-bit does frequently find errors, which is why it was readded recently.

Reply:
Sure, I mean find the 32-bit errors on Alpine, almost every API is the same as in Linux.

Reply:
Ah that's what you mean. I'm not sure how easy it is to build on alpine with 32 bit. I suspect it might not be easy.

Reply:
it's actually not just musl but also integration with various libs. We have got regular reports for OpenSSL: see the whole list of tests with issues on alpine: https://github.com/php/php-src/issues/14036 . I haven't checked if this PR runs the latest OpenSSL version but the plan would be to match the latest issues.

Is this actually still the only one using root for test execution? If so, then it is quite useful for FPM.

Unless we see some significant slow down, I don't think there is really reason to not have it on push.

Reply:
I see `Successful in 15m` which is faster than many other test so I can't really see why it would be an issue to include it. Those are low resource runs so energy argument is not exactly valid either. :)

Reply:
once it becomes the slowest build and it doesn't catch any new issues, we can for sure move. But so far it's one of the faster runs by the look of it so I think we should keep it on push. I think it has got difference which you can see from the linked OpenSSL issues and I have seen also difference in networking / threading / memory allocator. Also alpine docker images are quite popular so it's good to have it tested on push especially if it doesn't cost anything.

Reply:
Yes, the popularity of alpine docker images is the primary reason for which I think we should better support alpine/musl, which implies we should have automated testing for alpine/musl.

However I don't know what to do with the push job.

Since I initially opened this PR in April, the test suite was broken only 3 times. On two occasions it was just a broken test (https://github.com/php/php-src/pull/14790, https://github.com/php/php-src/pull/15072), but the last time it was a more serious regression (https://github.com/php/php-src/issues/15384), which ended up in beta1.

In theory a nightly job would be enough to prevent regressions from ending up in a release, as long as we check nightly job status before tagging, and we tag the commit used by the job. According to https://github.com/php/php-src/blob/master/docs/release-process.md, checking the tests status is already part of the process.

Reply:
Yeah but nightly is usually not checked for the release tag as we usually use the latest commit. I sometimes check nightly for the previous day before the release but not sure if other RM's do that. So the push is safer for release.

Reply:
You can trigger nightly manually:

https://github.com/php/php-src/actions/workflows/nightly.yml

Under "Run workflow". That will always be more valuable than putting more things into push.yml, as it will actually run ASAN on all extensions, which is likely the most valuable check anyway. It will also run MSAN, more opcache, tests, etc.

Reply:
We could also add an on-tag trigger for nightly.

Reply:
The on-tag is too late as we don't want to waste tag version ideally. Although for RC it might not be such a big issue really so might be good idea to extend that one. The only issue is that this commit is often not checked either but the future automation plan would make use of it.

Triggering manually is not ideal either because it adds waiting time so we don't really want to do that.

In any case I still haven't heard a proper reason why it shouldn't be part of push. The only reason was really that it might add up time but currently it doesn't so why not just add it and if there's a problem, then we can review it?

Reply:
There's not one specific point where this becomes a problem. But omitting any redundant jobs can save load on the planet and reduce CI queues. I don't object to adding a job that adds value, but the argumentation so far wasn't particularly convincing (i.e. one broken case in months, that would be uncovered the following day instead of immediately). Some suggestions:

1. Omit push for now, and reconsider if I notice that it fails more frequently than expected.
2. Add it in favor of some other job. E.g. LINUX_X64_RELEASE_NTS doesn't seem particularly crucial, if we add the release flag to some other existing test. If Alpine tests with asan, it's better not to add it here.

Reply:
This is not redundant. Have you seen the OpenSSL issue list: https://github.com/php/php-src/issues/14036 ? This already shows that there are significant enough differences. Catching issues after the merge usually takes time and it is also riskier for releases.

As I mentioned those are low resource jobs so the load on planet is minimal (you should really see what normally runs in enterprise pipelines where usually you don't see any nightly runs and lots of jobs are high resource ones - everything is often done on push).

The way how I see it is to have only long running and pipelines with not enough credits to be run in nightly. Anywya we don't probably agree on this so I will leave decision to you as you are the code owner for this part. Feel free to remove the push then.

Reply:
As mentioned, I don't inherently object to increasing CI time, but I do think we should keep it minimal, rather than keep growing it until it becomes a problem. I don't know what the OpenSSL issue is about, but since this change wasn't even caused by us, it wouldn't have prevented the failing of release branches either.

Anyway, this discussion has already gone on too long to be useful. I'll leave it up to Arnaud to decide whether to keep push or remove it.

Reply:
@arnaud-lb I understand @iluuu1994 as any added CI pipeline increases the CI time meaning not only more energy used but also CI might delay other CIs for different PRs.

However I understand @bukka as well. Push CI pipelines are great for evaluating if PR or any commit is working properly.

I was neutral on this because of these pros and cons, but in past I strugled with https://github.com/php/php-src/issues/13771 many times and as reported there, many times this was left unnoticed for a while on master branch.

Based on this, I would support having Alpine tested on push with https://github.com/php/php-src/issues/13771, ie. one push job where the extensions would be build separately testing if the header files and dependencies of each extension are defined properly on push.

Reply:
I've decided to merge only the nightly job for now. I will add a push job if nightly fails regularly.

@mvorisek I think what you suggest in #13771 would solve your issue as it would detect issues less than 24 hours after they were introduced?

Reply:
#13771 about every second month the build was broken, so I would say it should be tested on push as it is not only an Alpine issue, but a build issue in general